test/unit/puppetlabs/puppetserver/ruby/http_client_test.clj

```diff
@@ -176,59 +173,72 @@ jruby-config :- jruby-schemas/JRubyConfig
 (deftest test-ruby-http-client
   (jetty9/with-test-webserver ring-app port
     (with-scripting-container sc
-      (with-http-client sc port {:use-ssl false}
-      (testing "HTTP GET"
-        (is (= "hi" (.runScriptlet sc "$c.get('/', {}).body"))))
-      (testing "HTTP POST"
-        (is (= "hi" (.runScriptlet sc "$c.post('/', 'foo', {}).body"))))))))
+      (with-http-client sc {}
+        (let [url (str "http://localhost:" port)]
+          (testing "HTTP GET"
+            (is (= "hi" (.runScriptlet sc (format "$c.get(URI('%s')).body" url)))))
+          (testing "HTTP POST"
+            (is (= "hi" (.runScriptlet sc (format "$c.post(URI('%s'), 'foo').body" url))))))))))
+
+(deftest http-escaped-urls-test
+  (jetty9/with-test-webserver ring-app port
+    (with-scripting-container sc
+      (with-http-client sc {}
+        (let [url (str "http://localhost:" port)]
+          (testing "HTTP GET"
+            (.runScriptlet sc (str "$response = $c.get(URI('" url "/a%20b%3Fc'), params: { foo: 'bar' })"))
+            (is (= 200 (.runScriptlet sc "$response.code")))
+            (is (= (str url "/a%20b%3Fc?foo=bar") (.runScriptlet sc "$response.url.to_s")))))))))
 
 (deftest http-basic-auth
   (jetty9/with-test-webserver ring-app-with-auth port
     (with-scripting-container sc
-      (with-http-client sc port {:use-ssl false}
-        (testing "no credentials"
-          (.runScriptlet sc "$response = $c.post('/', 'foo', {})")
-          (is (= "401" (.runScriptlet sc "$response.code")))
-          (is (= "access denied" (.runScriptlet sc "$response.body"))))
-
-        (testing "valid credentials"
-          (let [auth "{ :basic_auth => { :user => 'foo', :password => 'bar' }}"]
-            (.runScriptlet sc (format "$response = $c.post('/', 'foo', {}, %s)" auth)))
-          (is (= "200" (.runScriptlet sc "$response.code")))
-          (is (= "hi" (.runScriptlet sc "$response.body"))))
-
-        (testing "invalid credentials"
-          (let [auth "{ :basic_auth => { :user => 'foo', :password => 'baz' }}"]
-            (.runScriptlet sc (format "$response = $c.post('/', 'foo', {}, %s)" auth)))
-          (is (= "401" (.runScriptlet sc "$response.code")))
-          (is (= "access denied" (.runScriptlet sc "$response.body"))))))))
+      (with-http-client sc {}
+        (let [url (str "http://localhost:" port)]
+          (testing "no credentials"
+            (.runScriptlet sc (format "$response = $c.post(URI('%s'), 'foo')" url))
+            (is (= 401 (.runScriptlet sc "$response.code")))
+            (is (= "Unauthorized" (.runScriptlet sc "$response.reason")))
+            (is (= "access denied" (.runScriptlet sc "$response.body"))))
+
+          (testing "valid credentials"
+            (let [auth "{ :basic_auth => { :user => 'foo', :password => 'bar' }}"]
```

Comment:
Are these credentials set up somewhere? I'm not noticing a difference between this test and the "invalid credentials" test, other than the password value. Just curious how this works.

Reply:
We set up an app with authenticated middleware [here](https://github.com/puppetlabs/puppetserver/blob/be1af42078c344d5f070ff36df7a70be33c1fb25/test/unit/puppetlabs/puppetserver/ruby/http_client_test.clj#L53-L59) and pass it as the app the macro that sets up the webserver at the top of this test.